Add check for existing query params

wizmo2 opened the pull request:
<!--
  You are amazing! Thanks for contributing to our project!
  Please, DO NOT DELETE ANY TEXT from this template! (unless instructed).
-->
## Breaking change
<!--
  If your PR contains a breaking change for existing users, it is important
  to tell them what breaks, how to make it work again and why we did this.
  This piece of text is published with the release notes, so it helps if you
  write it towards our users, not us.
  Note: Remove this section if this PR is NOT a breaking change.
-->
2020.2 introduced changes to Control users and Front-end that resulted in links for cached images in Media Browser having additional authSig parameters appended (for Custom components)

With the existing code, if the path already contains the query operator ('?') existing queries become corrupted.  For example, the cached images in MediaPlayerImageView have a token query that must be authorized.  In the current build, the authorization fails as the url is `/api/.......jpg?token=xxxxx?authSig=xxxxx`    

## Proposed change
<!--
  Describe the big picture of your changes here to communicate to the
  maintainers why we should accept this pull request. If it fixes a bug
  or resolves a feature request, be sure to link to that issue in the
  additional information section.
-->
Add simple check when adding authSig to existing path for existing query parameters


## Type of change
<!--
  What type of change does your PR introduce to Home Assistant?
  NOTE: Please, check only 1! box!
  If your PR requires multiple boxes to be checked, you'll most likely need to
  split it into multiple PRs. This makes things easier and faster to code review.
-->

- [ ] Dependency upgrade
- [x] Bugfix (non-breaking change which fixes an issue)
- [ ] New integration (thank you!)
- [ ] New feature (which adds functionality to an existing integration)
- [ ] Breaking change (fix/feature causing existing functionality to break)
- [ ] Code quality improvements to existing code or addition of tests

## Additional information
<!--
  Details are important, and help maintainers processing your PR.
  Please be sure to fill out additional details, if applicable.
-->

- This PR fixes or closes issue: fixes #
- This PR is related to issue: 
- Link to documentation pull request: 

## Checklist
<!--
  Put an `x` in the boxes that apply. You can also fill these out after
  creating the PR. If you're unsure about any of them, don't hesitate to ask.
  We're here to help! This is simply a reminder of what we are going to look
  for before merging your code.
-->

- [x] The code change is tested and works locally.
- [x] Local tests pass. **Your PR cannot be merged unless tests pass**
- [x] There is no commented out code in this PR.
- [x] I have followed the [development checklist][dev-checklist]
- [ ] The code has been formatted using Black (`black --fast homeassistant tests`)
- [ ] Tests have been added to verify that the new code works.

If user exposed functionality or configuration variables are added/changed:

- [ ] Documentation added/updated for [www.home-assistant.io][docs-repository]

If the code communicates with devices, web services, or third-party tools:

- [ ] The [manifest file][manifest-docs] has all fields filled out correctly.  
      Updated and included derived files by running: `python3 -m script.hassfest`.
- [ ] New or updated dependencies have been added to `requirements_all.txt`.  
      Updated by running `python3 -m script.gen_requirements_all`.
- [ ] For the updated dependencies - a link to the changelog, or at minimum a diff between library versions is added to the PR description.
- [ ] Untested files have been added to `.coveragerc`.

The integration reached or maintains the following [Integration Quality Scale][quality-scale]:
<!--
  The Integration Quality Scale scores an integration on the code quality
  and user experience. Each level of the quality scale consists of a list
  of requirements. We highly recommend getting your integration scored!
-->

- [ ] No score or internal
- [ ] 🥈 Silver
- [ ] 🥇 Gold
- [ ] 🏆 Platinum

<!--
  This project is very active and we have a high turnover of pull requests.

  Unfortunately, the number of incoming pull requests is higher than what our
  reviewers can review and merge so there is a long backlog of pull requests
  waiting for review. You can help here!
  
  By reviewing another pull request, you will help raise the code quality of
  that pull request and the final review will be faster. This way the general
  pace of pull request reviews will go up and your wait time will go down.
  
  When picking a pull request to review, try to choose one that hasn't yet
  been reviewed.

  Thanks for helping out!
-->

To help with the load of incoming pull requests:

- [ ] I have reviewed two other [open pull requests][prs] in this repository.

[prs]: https://github.com/home-assistant/core/pulls?q=is%3Aopen+is%3Apr+-author%3A%40me+-draft%3Atrue+-label%3Awaiting-for-upstream+sort%3Acreated-desc+review%3Anone+-status%3Afailure

<!--
  Thank you for contributing <3

  Below, some useful links you could explore:
-->
[dev-checklist]: https://developers.home-assistant.io/docs/en/development_checklist.html
[manifest-docs]: https://developers.home-assistant.io/docs/en/creating_integration_manifest.html
[quality-scale]: https://developers.home-assistant.io/docs/en/next/integration_quality_scale_index.html
[docs-repository]: https://github.com/home-assistant/home-assistant.io

wizmo2 committed:
Add check for existing query params

elupus commented:
Please fix this with a proper url parser. This is too fragile. Would break with fragments too.

wizmo2 commented:
Agreed, but this fixes the broken code for now.  

If we want to add full parse verification through out HA, we'll probably need to create a custom function to append queries as urllib doesn't do this out of the box

balloob commented on homeassistant/components/http/auth.py:
```diff
@@ -67,6 +67,8 @@ def async_sign_path(
         secret,
         algorithm="HS256",
     )
+    if "?" in path
```
- Use `yarl.URL` for mutating URLs
- We need to decide if the auth signature now also need to verify that the passed query params are like at the time of signing. 
- Tests, tests, tests

balloob commented:
For your custom component, you can also strip the `token` param because the `authSig` will provide the authentication.

wizmo2 commented:
That works, just had to use my own get_browse_image_url. Thanks

I'll still have a go at this PR.  Going to look at yarl as I see it has a update_query.

wizmo2 committed:
Add full url parse operation to append authSig

Added yarl, as this has the simplest implementation of query appends.  

Considered changing urllib unquote to use yarl also, but decided there was not specific need

wizmo2 commented:
Added yarl to update query parameters.

Did look at replacing urllib unquote, but didn't want to add additional risk

elupus commented on homeassistant/components/http/auth.py:
```diff
@@ -67,7 +68,7 @@ def async_sign_path(
         algorithm="HS256",
     )
-    return f"{path}?{SIGN_QUERY_PARAM}={encoded}"
+    return str(URL(path)%f"{SIGN_QUERY_PARAM}={encoded}")
```
Feels like this should be: `str(URL(path) % {SIGN_QUERY_PARAM: unencoded})`

balloob replied:
YARL has full support for updating query params. 

However, we need to encode the query params and make sure that those are exactly as-is when the request comes in.

wizmo2 commented:
May need help (or time) on the test setup.  Have not had a chance to look into how that works.  

I would like to get this resolved, as "token strip" solution creates a distinct HA version compatibility transition in my custom component code.
